### PR TITLE
feat: add video input connector and streaming endpoints

### PR DESCRIPTION
**Pull request by Bycob**
- [x] Unit tests
- [x] Usage example

### From a file

```bash
#!/bin/bash

# Create resource
curl -X PUT "http://localhost:8500/resources/video_file" -d '{
    "type": "video",
    "source": "/opt/platform/data/video.MOV"
}'

echo ""
curl -X PUT "http://localhost:8500/services/imageserv" -d '{
    "description": "image detection service",
    "mllib": "caffe",
    "model": {
        "repository": "/opt/platform/models/public/detection_600/"
    },
    "parameters": {
        "input": {
            "connector": "image",
            "width": 300,
            "height": 300
        },
        "mllib": {
            "nclasses": 601,
            "gpu": true,
            "gpuid": 1
        }
    },
    "type": "supervised"
}'
echo ""

running=1

while (( running == 1 ))
do
    # Infer on current frame and skil to next frame
    result=$(curl -X POST "http://localhost:8500/predict" -d '{
       "service":"imageserv",
       "parameters":{
         "input": {},
         "output":{
           "bbox": true,
           "confidence_threshold": 0.1
         }
       },
       "data":["video_file"]
    }')
    echo "${result}"
    status_code=$(jq -r '.status.code' <<< "${result}")
    res_status=$(jq -r '.body.resources[0].status' <<< "${result}")
    res_end=$([[ "$res_status" == "ended" ]] && echo 1 || echo 0)

    if (( status_code != 200 || res_end == 1 ))
    then
         running=0
    fi
done

curl -X DELETE "http://localhost:8080/resources/video_file"
```

### From a youtube video

Exactly the same thing except for resource creation call. Here we are using gstreamer to read a video from the internet.

```bash
# Create resource
curl -X PUT "http://localhost:8500/resources/video_youtube" -d '{                                                                                                                                          
    "type": "video",                                                                                                                                                                                       
    "source": "souphttpsrc is-live=true location=\"'$(youtube-dl --format "best[ext=mp4][protocol=https]" --get-url "https://youtu.be/J---aiyznGQ")'\" ! qtdemux ! h264parse ! avdec_h264 ! videoconvert ! 
appsink"                                                                                                                                                                                                   
}'

echo ""
curl -X PUT "http://localhost:8500/services/imageserv" -d '{
    "description": "image detection service",
    "mllib": "caffe",
    "model": {
        "repository": "/opt/platform/models/public/detection_600/"
    },
    "parameters": {
        "input": {
            "connector": "image",
            "width": 300,
            "height": 300
        },
        "mllib": {
            "nclasses": 601,
            "gpu": true,
            "gpuid": 1
        }
    },
    "type": "supervised"
}'
echo ""

running=1

while (( running == 1 ))
do
    # Infer on current frame and skil to next frame
    result=$(curl -X POST "http://localhost:8500/predict" -d '{
       "service":"imageserv",
       "parameters":{
         "input": {},
         "output":{
           "bbox": true,
           "confidence_threshold": 0.1
         }
       },
       "data":["video_youtube"]
    }')
    echo "${result}"
    status_code=$(jq -r '.status.code' <<< "${result}")
    res_status=$(jq -r '.body.resources[0].status' <<< "${result}")
    res_end=$([[ "$res_status" == "ended" ]] && echo 1 || echo 0)

    if (( status_code != 200 || res_end == 1 ))
    then
         running=0
    fi
done
```

### CODE REVIEW

File: src/http/controller.hpp [hunk 1]
```diff
@@ -231,6 +233,76 @@ class DedeController : public oatpp::web::server::api::ApiController
...
+                             _oja->get_stream_info(stream_name->std_str()));
+  }
+
+  ENDPOINT_INFO(delete_stream)
+  {
+    info->summary = "Stop and remove a running stream";
+    info->addResponse<Object<dd::DTO::GenericResponse>>(Status::CODE_200,
+                                                        "application/json");
+  }
+  ENDPOINT("DELETE", "streams/{stream-name}", delete_stream,
```
Comment (beniz) [inline src/http/controller.hpp:300]: `/stream` instead of `/streams` ?

**Comment by beniz** [comment]
```
/data1/jenkins/workspace/deepdetect_PR-1375/src/resources.cc: In static member function 'static dd::res_variant_type dd::ResourceFactory::create(const string&, oatpp::Object<dd::DTO::Resource>&)':

/data1/jenkins/workspace/deepdetect_PR-1375/src/resources.cc:227:25: error: redundant move in return statement [-Werror=redundant-move]

  227 |         return std::move(res_variant);

      |                ~~~~~~~~~^~~~~~~~~~~~~
```
from the CI

File: src/resources.h [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    const char *what() const noexcept
+    {
+      return _s.c_str();
+    }
+
+  private:
+    std::string _s;
+  };
+
+  // ====
```
Comment (beniz) [inline src/resources.h:117]: Remove ?

File: src/resources.cc [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+ */
+
+#include "resources.h"
+
+#include <boost/algorithm/string/predicate.hpp>
+
+#include "utils/cv_utils.hpp"
+
+#ifndef DD_FOURCC_MACRO
+#define DD_FOURCC_MACRO(c1, c2, c3, c4)                                       \
```
Comment (beniz) [inline src/resources.cc:29]: Is this used somewhere ?

File: src/resources.cc [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        throw ResourceBadParamException("Video with uri \"" + uri
+                                        + "\" could not be opened");
+      }
+
+    int fps = _capture.get(cv::CAP_PROP_FPS);
+    cv::Size dims((int)_capture.get(cv::CAP_PROP_FRAME_WIDTH),
+                  (int)_capture.get(cv::CAP_PROP_FRAME_HEIGHT));
+    auto frame_count = uint32_t(_capture.get(cv::CAP_PROP_FRAME_COUNT));
+    int fourcc = static_cast<int>(_capture.get(cv::CAP_PROP_FOURCC));
+    this->_logger->info("Video properties: {}x{} - {} fps, {} frames, enc={}",
```
Comment (beniz) [inline src/resources.cc:172]: Are we returning this info as JSON output to the API call ? If not, could be useful to do so!

File: src/resources.cc (hunk 3 above)
Comment (Bycob) [reply src/resources.cc:29]: Removed

File: src/resources.cc (hunk 4 above)
Comment (Bycob) [reply src/resources.cc:167]: Done

**Review by beniz — APPROVED** [review]

**Review by fantes — APPROVED** [review]

**Pull request merged**